std/java/_std/sys/thread/Thread.hx

```diff
@@ -23,89 +23,266 @@
-	public static function readMessage(block:Bool):Dynamic {
-		return current().getHandle().messages.pop(block);
+	public static inline function readMessage(block:Bool):Dynamic {
+		return current().getHandle().readMessage(block);
 	}
 
 	public inline function sendMessage(msg:Dynamic):Void {
 		this.sendMessage(msg);
 	}
 
-	private inline function getHandle():NativeThread {
+	public static inline function repeatEvent(event:()->Void, intervalMs:Int):EventHandler {
+		return current().getHandle().repeatEvent(event, intervalMs);
+	}
+
+	public static inline function cancelEvent(eventHandler:EventHandler):Void {
+		current().getHandle().cancelEvent(eventHandler);
+	}
+
+	public inline function runEvent(event:()->Void):Void {
+		this.runEvent(event);
+	}
+
+	public inline function runPromisedEvent(event:()->Void):Void {
+		this.runPromisedEvent(event);
+	}
+
+	public inline function promiseEvent():Void {
+		this.promiseEvent();
+	}
+
+	inline function getHandle():HaxeThread {
+		return this;
+	}
+
+	@:keep
+	private static function processEvents():Void {
+		current().getHandle().processEvents();
+	}
+}
+
+abstract EventHandler(RegularEvent) from RegularEvent {
+	inline function get():RegularEvent {
 		return this;
 	}
 }
 
-@:native('haxe.java.vm.Thread') private class NativeThread {
-	@:private static var javaThreadToHaxe = new haxe.ds.WeakMap<java.lang.Thread, NativeThread>();
-	@:private static var mainJavaThread = java.lang.Thread.currentThread();
-	@:private static var mainHaxeThread = {
-		var ret = new NativeThread();
-		javaThreadToHaxe.set(mainJavaThread, ret);
-		ret;
-	};
-
-	public static function getThread(jt:java.lang.Thread):NativeThread {
-		if (Std.isOfType(jt, HaxeThread)) {
-			var t:HaxeThread = cast jt;
-			return t.threadObject;
-		} else if (jt == mainJavaThread) {
+private class HaxeThread {
+	static final nativeThreads = Collections.synchronizedMap(new WeakHashMap<JavaThread,HaxeThread>());
+	static final mainJavaThread = JavaThread.currentThread();
+	static final mainHaxeThread = new HaxeThread();
+
+	public final messages = new LinkedBlockingDeque<Dynamic>();
+	final oneTimeEvents = new LinkedBlockingDeque<()->Void>();
+	final promisedEvents = new AtomicInteger();
+	var regularEvents:Null<RegularEvent>;
+
+	public static function create(job:()->Void):HaxeThread {
+		var hx = new HaxeThread();
+		var thread = new NativeHaxeThread(hx, job);
+		thread.setDaemon(true);
+		thread.start();
+		return hx;
+	}
+
+	public static function get(javaThread:JavaThread):HaxeThread {
+		if(javaThread == mainJavaThread) {
 			return mainHaxeThread;
+		} else if(javaThread is NativeHaxeThread) {
+			return (cast javaThread:NativeHaxeThread).haxeThread;
 		} else {
-			var ret = null;
-			untyped __lock__(javaThreadToHaxe, {
-				ret = javaThreadToHaxe.get(jt);
-				if (ret == null) {
-					ret = new NativeThread();
-					javaThreadToHaxe.set(jt, ret);
+			switch nativeThreads.get(javaThread) {
+				case null:
+					var hx = new HaxeThread();
```

Comment:
This does not call processEvents, meaning stuff dispatched from this thread would be skipped silently. Here's something that would work:

```haxe
var waitAndProcess:java.lang.Runnable = cast function () {
  javaThread.join();
  nativeThreads.put(JavaThread.currentThread(), hx);
  hx.processEvents();
}
var eventProcessor = new java.lang.Thread(waitAndProcess:java);
eventProcessor.setDaemon(true);
eventProcessor.start();
```

Still, if the calling thread itself is looped (e.g. because it's from some library that has its own even loop), then this also won't work.

The alternatives I see:

- don't "promote" native threads, either throwing when `Thread.current()` is called from a non-haxe-created thread 
or change the signature of `Thread.current()` to return `Null<Thread>` on java
- have a fallback event loop ... adding events to it will be a bit trickier, because it needs to be thread safe and may have to alter the thread status, i.e. if its thread is terminated it needs to be run again, and if it's currently waiting for a timeout then you have to interrupt it ... and assuming all that works, you'll still lose the guarantee that callbacks land on the same thread

Reply:
The idea is that for non-haxe threads users have to decide when to run event loop manually with `@:privateAccess Thread.processEvents();`. I think it's ok to make it private (with an appropriate doc) because it's an advanced feature and we don't want users to manually start event loops on haxe-threads.

Throwing from `Thread.current()` may be better indeed, but then we should provide an ability to instrument non-haxe threads with haxe event loops.

Reply:
> Throwing from `Thread.current()` may be better indeed, but then we should provide an ability to instrument non-haxe threads with haxe event loops.

Not so sure about that. It's always possible that the thread in question is looping infinitely by design. Could be a worker from some async IO library, or worst case it could be the UI thread. Perhaps the user would be better off transferring control to a haxe thread (via message or one time task). Or we could throw this:

```haxe
class NonHaxeThreadException extends Exception {
  public final javaThread:java.lang.Thread;// for inspection
  public function promoteThread(scheduleEvents:(processEvents:(ms:Int)->Void)->Void):sys.thread.Thread;
}
```

In that case the user can attempt to schedule the loop via the mechanisms provided by the library that created the thread.

There's also the case that the user is simply creating a Timer on a non-haxe thread. Would be annoying to have to try/catch-wrap all thread creation. Perhaps there could be a `static dynamic` function exposed on `Timer` or `Thread` to allow users creating a fallback behavior that fits their use case best?

This is tricky. But in any case, I think ultimately it's the right move. It's definitely better than dispatching everything to the main thread, which may already be terminated.